docs/adr/0006-consensus-governance.md

```diff
@@ -0,0 +1,478 @@
+publishing an upgrade descriptor or cancelling a previously passed upgrade.
+
+### State
+
+#### Staking
+
+This proposal adds the following consensus layer state in the staking module:
+
+- **Governance deposits account balance (`0x59`)**, similar to the common pool.
+
+#### Governance
+
+This proposal adds the following consensus layer state in the governance module:
+
+- **Next proposal identifier (`0x80`)**
+
+  The next proposal identifier is stored as a CBOR-serialized `uint64`.
+
+- **List of proposals (`0x81`)**
+
+  Each proposal is stored under a separate storage key with the following key
+  format:
+
+  ```
+  0x81 <proposal-id (uint64)>
+  ```
+
+  And CBOR-serialized value:
+
+  ```golang
+  // ProposalState is the state of the proposal.
+  type ProposalState uint8
+
+  const (
+      StateActive   ProposalState = 1
+      StatePassed   ProposalState = 2
+      StateRejected ProposalState = 3
+      StateFailed   ProposalState = 4
+  )
+
+  type Proposal struct {
+      // ID is the unique identifier of the proposal.
+      ID uint64 `json:"id"`
+      // Submitter is the address of the proposal submitter.
+      Submitter staking.Address `json:"submitter"`
+      // State is the state of the proposal.
+      State ProposalState `json:"state"`
+      // Deposit is the deposit attached to the proposal.
+      Deposit quantity.Quantity `json:"deposit"`
+
+      // Content is the content of the proposal.
+      Content ProposalContent `json:"content"`
+
+      // CreatedAt is the epoch at which the proposal was created.
+      CreatedAt beacon.EpochTime `json:"created_at"`
+      // ClosesAt is the epoch at which the proposal will close and votes will
+      // be tallied.
+      ClosesAt beacon.EpochTime `json:"closes_at"`
+
+      // Results are the final tallied results after the voting period has
+      // ended. Note that they take the
+      Results map[Vote]quantity.Quantity `json:"results,omitempty"`
+      // InvalidVotes is the number of invalid votes after tallying.
+      InvalidVotes uint64 `json:"invalid_votes,omitempty"`
+  }
+  ```
+
+- **List of active proposals (`0x82`)**
+
+  Each active proposal (one that has not yet closed) is stored under a separate
+  storage key with the following key format:
+
+  ```
+  0x82 <closes-at (beacon.EpochTime)> <proposal-id (uint64)>
+  ```
+
+  The value is empty as the proposal ID can be inferred from the key.
+
+- **List of votes (`0x83`)**
+
+  Each vote is stored under a separate storage key with the following key
+  format:
+
+  ```
+  0x83 <proposal-id (uint64)> <voter-address (staking.Address)>
+  ```
+
+  And CBOR-serialized value:
+
+  ```golang
+  // Vote is a governance vote.
+  type Vote uint8
+
+  const (
+      VoteYes     Vote = 1
+      VoteNo      Vote = 2
+      VoteAbstain Vote = 3
+  )
+  ```
+
+- **List of pending upgrades (`0x84`)**
+
+  Each pending upgrade is stored under a separate storage key with the following
+  key format:
+
```

Comment:
I assume a CBOR-serialized `UpgradeProposal` should be stored here

Reply:
No need, the idea is that you just need to get the ID and then you can fetch the proposal from `0x81 ...`.

Reply:
Yeah makes sense :+1: